certora/specs/RewardsDistributor.spec

```diff
@@ -0,0 +1,67 @@
+using MerkleTrees as MerkleTrees;
+using MorphoToken as MorphoToken;
+
+methods {
+    function prevRoot() external returns bytes32 envfree;
+    function currRoot() external returns bytes32 envfree;
+    function claimed(address) external returns uint256 envfree;
+    function claim(address, uint256, bytes32[]) external envfree;
+
+    function MerkleTrees.getValue(address, address) external returns uint256 envfree;
+    function MerkleTrees.getHash(address, address) external returns bytes32 envfree;
+    function MerkleTrees.wellFormedUpTo(address, address, uint256) external envfree;
+
+    function MorphoToken.balanceOf(address) external returns uint256 envfree;
+}
+
+// Check an account can only claim greater rewards each time.
+rule noClaimAgain(address _account, uint256 _claimable, bytes32[] _proof) {
+    claim(_account, _claimable, _proof);
+
+    assert _claimable == claimed(_account);
+
+    uint256 _claimable2;
+    // Assume that the second claim is smaller or equal to the previous claimed amount.
+    require (_claimable2 <= _claimable);
+    claim@withrevert(_account, _claimable2, _proof);
+
+    assert lastReverted;
+}
+
+// Check that the transferred amount is equal to the claimed amount minus the previous claimed amount.
+rule transferredTokens(address _account, uint256 _claimable, bytes32[] _proof) {
+    // Assume that the rewards distributor itself is not receiving the tokens, to simplify this rule.
+    require _account != currentContract;
+
+    uint256 balanceBefore = MorphoToken.balanceOf(_account);
+    uint256 claimedBefore = claimed(_account);
+
+    // Safe require because the sum is capped by the total supply.
+    require balanceBefore + MorphoToken.balanceOf(currentContract) < 2^256;
+
+    claim(_account, _claimable, _proof);
+
+    uint256 balanceAfter = MorphoToken.balanceOf(_account);
+
+    assert balanceAfter - balanceBefore == _claimable - claimedBefore;
+}
+
```

Comment:
I find this rule a bit odd because it is assumed that there should be only one leaf in the Merkle tree for each address (this is actually checked in the rewards checker [here](https://github.com/morpho-org/morpho-optimizers/blob/ac2af4cc1889ed8b138460b8644a7c40c81a541a/certora/helpers/MerkleTreeLib.sol#L34)). So this means that you should not be able to claim twice, for the same root

Reply:
Actually since we have 2 roots it would not be absurd to check this. Those 2 roots are here precisely for that reason of being able to claim even if the root update frontruns your claim transactions. Let me try it

Reply:
Done in [48fb135](https://github.com/morpho-org/morpho-optimizers/pull/1656/commits/48fb135fc4a6cc39a1e76d7a955b82c502f05f3f)